### PR TITLE
[Diagnostics] Diagnose unresolved and typoed labels when using break/continue inside loops

### PR DESCRIPTION
**Pull request by robinkunde**
Adds proper diagnostics when using unresolved labels on break statements inside loops. In case of simple typos, also provides suggestions and fixits. Right now, the misleading "error: 'break' is only allowed inside a loop, if, do, or switch" is emitted in 4.2 and master.

Resolves [SR-9677](https://bugs.swift.org/browse/SR-9677).

Note: I made an effort to use code/constants from the typechecker. Let me know if my edits to TypeChecker.h need to be redone.

### CODE REVIEW

File: lib/Sema/TypeCheckNameLookup.cpp [hunk 1]
```diff
@@ -574,16 +574,7 @@ LookupResult TypeChecker::lookupConstructors(DeclContext *dc, Type type,
...
-  UnreasonableCallEditDistance = 8,
-
-  /// Don't consider candidates that score worse than the given distance
-  /// from the best candidate.
-  MaxCallEditDistanceFromBestCandidate = 1
-};
-
-static unsigned getCallEditDistance(DeclName writtenName,
+unsigned TypeChecker::getCallEditDistance(DeclName writtenName,
                                     DeclName correctedName,
```
Comment (slavapestov) [inline lib/Sema/TypeCheckNameLookup.cpp]: Nitpick: indentation should line up with (. Also, can this be a static method? It doesn't use any per-instance state from TypeChecker

**Review by slavapestov — APPROVED** [review]
Thank you!

File: lib/Sema/TypeCheckNameLookup.cpp (hunk 1 above)
Comment (robinkunde) [reply lib/Sema/TypeCheckNameLookup.cpp]: Thank you for the feedback. I've amended the commit to fix indentation and make it static.

**Comment by jrose-apple** [comment]
I trust Slava's review, but can you get `continue label` at the same time? :-)

**Comment by robinkunde** [comment]
@jrose-apple You mean as part of this PR, right? I can do that.

**Comment by robinkunde** [comment]
@jrose-apple continue is now covered as well.

File: lib/Sema/TypeCheckStmt.cpp [hunk 2]
```diff
@@ -806,31 +808,54 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
...
+          // Otherwise produce a generic error.
+          TC.diagnose(S->getLoc(), diag::break_outside_loop);
+        }
+      } else {
+        // If an unresolved label was used, but we have a single correction,
+        // produce the specific diagnostic and fixit.
+        if (labelCorrections.size() == 1) {
+          for (auto &entry : labelCorrections)
+            TC.diagnose(S->getTargetLoc(), diag::unresolved_label_corrected,
+                        S->getTargetName(), (*entry.Value).getLabelInfo().Name)
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckStmt.cpp]: `entry.Value->getLabelInfo().Name`, yeah?

File: lib/Sema/TypeCheckStmt.cpp [hunk 3]
```diff
@@ -806,31 +808,54 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
...
+              .fixItReplace(SourceRange(S->getTargetLoc()),
+                            (*entry.Value).getLabelInfo().Name.get());
+        } else {
+          // If we have multiple corrections or none, produce a generic diagnostic
+          // and all fixits available.
+          auto diag = TC.diagnose(S->getTargetLoc(), diag::unresolved_label,
+                                  S->getTargetName());
+          for (auto &entry : labelCorrections)
+            diag.fixItReplace(SourceRange(S->getTargetLoc()),
+                              (*entry.Value).getLabelInfo().Name.get());
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckStmt.cpp]: Fix-its aren't alternatives; they're *all* applied. You'll have to either use multiple notes (in which case you could point at the statement that defines each label), or just string-append all the names into the diagnostic. (Or use a clever `%select` to put multiple variations in one diagnostic: "'foo'", "'foo' or 'bar'", "'foo', 'bar', or 'baz'".)

File: lib/Sema/TypeCheckStmt.cpp [hunk 4]
```diff
@@ -858,28 +884,53 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
...
+      labelCorrections.filterMaxScoreRange(
+        TypeChecker::MaxCallEditDistanceFromBestCandidate);
     }
 
-    if (!Target) {
+    if (Target) {
+      // Continue cannot be used to repeat switches, use fallthrough instead.
+      if (!Target->isPossibleContinueTarget()) {
+        TC.diagnose(S->getLoc(), diag::continue_not_in_this_stmt,
+                    isa<SwitchStmt>(Target) ? "switch" : "if");
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckStmt.cpp]: Oof. This should also apply to `guard` and `do`. I'll file a separate bug for that.

**Review by jrose-apple — COMMENTED** [review]
Thanks! It does seem like a lot of the `break` and `continue` logic could be shared, but I don't think that refactoring is a requirement for this PR to land. I did spot a place where the fix-its aren't going to work, though.

File: lib/Sema/TypeCheckStmt.cpp (hunk 2 above)
Comment (robinkunde) [reply lib/Sema/TypeCheckStmt.cpp]: Fixed.

File: lib/Sema/TypeCheckStmt.cpp (hunk 3 above)
Comment (robinkunde) [reply lib/Sema/TypeCheckStmt.cpp]: Yes, I tried to model on identifier typo correction, but I clearly misread that code on my first pass. Thank you for pointing this out.
I also refactored the shared logic into a static method.

File: lib/Sema/TypeCheckStmt.cpp [hunk 5]
```diff
@@ -858,31 +869,68 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
...
+    // produce the specific diagnostic and fixit.
+    if (corrections.size() == 1) {
+      tc.diagnose(targetLoc, diag::unresolved_label_corrected,
+                  targetName, corrections.begin()->Value->getLabelInfo().Name)
+        .highlight(SourceRange(targetLoc));
+      tc.diagnose(corrections.begin()->Value->getLabelInfo().Loc,
+                  diag::identifier_declared_here,
+                  corrections.begin()->Value->getLabelInfo().Name)
+        .fixItReplace(SourceRange(targetLoc),
+                      corrections.begin()->Value->getLabelInfo().Name.str());
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckStmt.cpp]: In this case I think it's reasonable to put the fix-it directly on the main diagnostic, which implies that it's the one obvious right thing to do.

**Comment by jrose-apple** [comment]
Thanks, this definitely looks better.

File: lib/Sema/TypeCheckStmt.cpp (hunk 5 above)
Comment (robinkunde) [reply lib/Sema/TypeCheckStmt.cpp]: Fixed

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
Thanks, Robin!

@swift-ci Please test

**Pull request merged**